cf-keycrypt/cf-keycrypt.c

```diff
@@ -0,0 +1,388 @@
+    "  -o       File to write encrypted/decrypted contents to. '-' writes to stdout.\n"
+    "  -H       Hostname/IP to encrypt for, gets data from lastseen database\n"
+    "  -h       Print help.\n"
+    "\n"
+    "Examples:\n"
+    "  Encrypt:\n"
+    "    cf-keycrypt -e /var/cfengine/ppkeys/localhost.pub -i myplain.dat -o mycrypt.dat\n"
+    "\n"
+    "  Decrypt:\n"
+    "    cf-keycrypt -d /var/cfengine/ppkeys/localhost.priv -i mycrypt.dat -o myplain.dat\n"
+    "\n"
+    "Written and maintained by Jon Henrik Bjornstad <jonhenrik@cfengineers.net>\n"
+    "\n"
+    "Copyright (C) cfengineers.net\n"
+    "\n"
+    );
+}
+
+void *get_in_addr(struct sockaddr *sa)
+{
+    if (sa->sa_family == AF_INET)
+    {
+        return &(((struct sockaddr_in*)sa)->sin_addr);
+    }
+    return &(((struct sockaddr_in6*)sa)->sin6_addr);
+}
+
+int file_exist(char *filename)
+{
+    struct stat buffer;
+    return (stat (filename, &buffer) == 0);
+}
+
+char *get_host_pubkey(char *host)
+{
+    // char *key;                            // TODO: this was unused(!)
+    // char value[BUFSIZE];                  // TODO: this was unused(!)
+    char *buffer = (char *) malloc(BUFSIZE *sizeof(char));
+    //char *keyname = NULL;
+    char hash[CF_HOSTKEY_STRING_SIZE];
+    //char buffer[BUFSIZE];
+    char ipaddress[BUFSIZE];
+    // int ecode;                            // TODO: this was unused(!)
+    struct addrinfo *result;
+    struct addrinfo *res;
+    bool found = false;
+
+    int error = getaddrinfo(host, NULL, NULL, &result);
+    if (error != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to get IP from host (getaddrinfo: %s)",
+            gai_strerror(error));
+        return NULL;
+    }
+
+    for(res = result; res != NULL && !found; res = res->ai_next)
+    {
+        inet_ntop(res->ai_family, get_in_addr((struct sockaddr *)res->ai_addr), ipaddress, sizeof(ipaddress));
+        if ((strcmp(ipaddress, "127.0.0.1") == 0) || (strcmp(ipaddress, "::1") == 0))
+        {
+            found = true;
+            snprintf(buffer, BUFSIZE * sizeof(char), "%s/ppkeys/localhost.pub", WORKDIR);
+            return buffer;
+        }
+        found = Address2Hostkey(hash, sizeof(hash), ipaddress);
+    }
+    if (found)
+    {
+        snprintf(buffer, BUFSIZE * sizeof(char), "%s/ppkeys/root-%s.pub", WORKDIR, hash);
+        return buffer;
+    }
+    else
+    {
+        for(res = result; res != NULL; res = res->ai_next)
+        {
+            inet_ntop(res->ai_family, get_in_addr((struct sockaddr *)res->ai_addr), ipaddress, sizeof(ipaddress));
+            snprintf(buffer, BUFSIZE * sizeof(char), "%s/ppkeys/root-%s.pub", WORKDIR, ipaddress);
+            if (file_exist(buffer))
+            {
+                return buffer;
+            }
+        }
+    }
+    return NULL;
+}
+
+void *readseckey(char *secfile)
+{
+    FILE *fp = NULL;
+    RSA* PRIVKEY = NULL;
+    static char *passphrase = "Cfengine passphrase";
```

Comment:
This exists all over our codebase. Move all separate definitions to `cf3.defs.h`. Consolidate this `readseckey()` function together with the existing code in crypto.c, export the one from there and import it here.

Reply:
I would rather not touch `crypto` or `defs`in this PR/commit. I can do that as a separate step after, if you'd like.